src/randomPicker.ts

```diff
@@ -41,6 +41,46 @@ export class RandomPicker<T> {
     return items;
   }
 
+  /**
+   * picks distinct elements from the weighted array.
+   * If the number of items required is more or equal to the length of weighted array, it will return all elements in array
+   * @param amount - positive number of items to be picked, if more or equal to the length of the array, returns all items
+   * @returns array of picked items
+   * @example
+   * const items = [
+   * { original: "Bronze", weight: 20 },
+   * { original: "Silver", weight: 10 },
+   * { original: "Gold", weight: 3 },
+   * { original: "Platinum", weight: 1 },
+   * ];
+   * const pickerAlwaysLast = new RandomPicker(items);
+   * const picked1 = pickerAlwaysLast.pickManyDistinct(1);
+   * console.log(picked1.length) // 1
+   * const picked3 = pickerAlwaysLast.pickManyDistinct(3);
+   * console.log(picked3.length) // 3
+   * const picked4 = pickerAlwaysLast.pickManyDistinct(4);
+   * console.log(picked4.length) // 4
+   * console.log(picked4) // ["Bronze", "Silver", "Gold", "Platinum"]
+   * const picked8 = pickerAlwaysLast.pickManyDistinct(8);
+   * console.log(picked8.length) // 4
+   * console.log(picked8) // ["Bronze", "Silver", "Gold", "Platinum"]
+   */
+  pickManyDistinct(amount: number): T[] {
+    if (amount < 0)
+      throw new Error(
+        "number of items to be picked should be a positive integer"
+      );
+    if (amount >= this.items.length) return this.items.map((i) => i.original);
+    const items: T[] = [];
+    const copyOfItems = [...this.items];
+    for (let i = 0; i < amount; i++) {
+      const indexToPick = this.indexToPick(copyOfItems);
+      items.push(copyOfItems[indexToPick].original);
+      copyOfItems.splice(indexToPick, 1);
```

Comment:
Instead of removing the item from `copyOfItems`, you can directly call `this.set(this.items.filter(i => i.original != picked ))` to remove the item from `RandomPicker`.

Outside the loop, before returning the data, you can restore with `this.set(copyOfItems)` to get back to the original state.